EIPS/eip-4361.md

```diff
@@ -0,0 +1,295 @@
+
+## Specification
+Sign-In with Ethereum works as follows:
+
+1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.
+2. The signature is then presented to the server, which checks the signature's validity and message content.
+3. Additional fields, including `expiration-time`, `not-before`, `request-id`, `chain-id`, and `resources` may be incorporated as part of authentication for the session.
+4. The server may further fetch data associated with the Ethereum address, such as from the Ethereum blockchain (e.g., ENS, account balances, ERC-20/ERC-721/ERC-1155 asset ownership), or other data sources that may or may not be permissioned.
+
+
+### Example message to be signed
+```
+service.org wants you to sign in with your Ethereum account:
+0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
+
+I accept the ServiceOrg Terms of Service: https://service.org/tos
+
+URI: https://service.org/login
+Version: 1
+Chain ID: 1
+Nonce: 328917
+Issued At: 2021-09-30T16:25:24Z
+Resources:
+- ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/
+- https://example.com/my-web2-claim.json
+```
+
+### Informal Message Template
+A Bash-like informal template of the full message is presented below for readability and ease of understanding. Field descriptions are provided in the following section. A full ABNF description is provided in the section thereafter.
+```
+${domain} wants you to sign in with your Ethereum account:
+${address}
+
+${statement}
+
+URI: ${uri}
+Version: ${version}
+Chain ID: ${chain-id}
+Nonce: ${nonce}
+Issued At: ${issued-at}
+Expiration Time: ${expiration-time}
+Not Before: ${not-before}
+Request ID: ${request-id}
+Resources:
+- ${resources[0]}
+- ${resources[1]}
+...
+- ${resources[n]}
+```
+### Message Field Descriptions
+- `domain` is the RFC 4501 `dnsauthority` that is requesting the signing.
+- `address` is the Ethereum address performing the signing conformant to capitalization encoded checksum specified in EIP-55 where applicable.
+- `statement` (optional) is a human-readable ASCII assertion that the user will sign, and it must not contain `'\n'`.
+- `uri` is an RFC 3986 URI referring to the resource that is the subject of the signing (as in the _subject of a claim_).
+- `version` is the current version of the `message`, which MUST be `1` for this specification.
+- `chain-id` is the EIP-155 Chain ID to which the session is bound, and the network where Contract Accounts must be resolved.
```

Comment:
Why is the `chain-id` not an optional parameter? Also why is the session being bound to chainId to start with?

What does `resolved` in thus context mean for Contract Accounts?

Reply:
Yes, it would be required for EIP-1271 based signing. Furthermore, by principle of least privilege, we wanted to strongly push implementers to lock the session to a specific `chain-id` or think carefully about its interpretation otherwise. I think keeping a required `chain-id` reduces one conditional check (for smart contract wallet or EOA) and also achieves the latter goal. Does that make sense?

Reply:
The flip side of this is that it requires me to re-authenticate each time I change networks in my wallet with the current UX flows. E.g. say I want to use sushiswap on mainnet and then decide to use a L2 solution I'd have to re-authenticate after changing networks which doesn't make sense to me. Wouldn't this lead to greater consent fatigue which is already pretty high as is? Furthermore, is it likely that we'd have the same account on different networks owned by different people?

Reply:
@kdenhartog An implementer is free to interpret a `chain-id` of `0` as across all networks for example, but that's their decision to make. From today's application, it does appear that networks aiming for EVM-compat indeed re-use the same account. The privacy and security implications of this are worth noting but out of scope for this spec.

Reply:
It sounds like it's being made in scope though because the chain-id is being required so that's a bit confusing to me. Requiring it's usage, but not stating why it's necessary nor placing additional normative requirements on it seems a bit odd to me and is opening the door for potential interoperability issues based on wallets and dApps making different assumptions.

Reply:
Would you onboard with adding the recommendation of `chain-id` to be `0` if it's a "don't care"?

Reply:
Yeah I think normatively defining that would be a good way to prevent interop concerns. Strong +1 from me. Would chain-id 0 also need to be registered with EIP 115?

Reply:
Hm, I don’t think it’s normatively defined in EIP-155, could be better to add the character `*` as a valid atom for this to avoid any future woes, okay with this?

Reply:
It also begs the question of smart contract wallets being able to sign cross chain…in earlier iterations we had a list of `chain-id` but found that to be overkill…

Reply:
I'm good with a `*` character, but I think it may encounter some weird logic in the way EIP-115 was originally intended, but I'm not certain there's overlap at this point. Particularly this clause in EIP-115 makes me think it could break some logic because it's not a number type:

> If block.number >= FORK_BLKNUM and v = CHAIN_ID * 2 + 35 or v = CHAIN_ID * 2 + 36


>It also begs the question of smart contract wallets being able to sign cross chain…in earlier iterations we had a list of chain-id but found that to be overkill…

Yeah that seems like it would be a really useful feature, but I'd assume based on your initial response that including this may be more scope creep than you'll want to take on at this point. I'm good with leaving the cross contract signing off for now in order to simplify, but there's definitely tradeoffs that will come with this in terms of limited capabilities here.

Reply:
> Would you onboard with adding the recommendation of `chain-id` to be `0` if it's a "don't care"?

Doesn't that imply `chain-id` is optional and the presumed default is 0? So no need to redeclare `0` with `0`.

Reply:
@kamescg I propose we keep it due to the reasoning here: https://github.com/ethereum/EIPs/pull/4361#discussion_r755524905

We can continue this discussion in the ETH Magicians forum if you feel strongly about it.